Add adUDotDot and infrastructure

cticenhour opened the pull request:
Closes #18178

cticenhour committed:
Implement ad_u_dot_dot (#18178)

lindsayad committed:
Switch to using references (#18178)

cticenhour committed:
Add AD second time derivative getters (#18178)

cticenhour committed:
Partially revert change to newmark-beta helper method (#18178)

cticenhour committed:
Add adCoupledDotDot method (#18178)

cticenhour committed:
Add tests for newmark beta and central difference (#18178)

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/766374/) on 69d56a2 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/18179/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/18179/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format db101a1bad37338030aabeef2ee91317bbee5e49`

cticenhour committed:
clang format (#18178)

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/766415/) on f7e5185 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/18179/site)

This comment will be updated on new commits.

cticenhour commented:
FYI - GitHub testing status did not update properly. All tests are green: https://civet.inl.gov/event/70202/

cticenhour commented:
@rwcarlsen Could you give this a review when you get a chance? Apologies for the potential spam - it doesn't have you listed as a "participant" yet, even though it tagged you as a Code Owner.

cticenhour committed:
Implement adUDotDot in MooseVariableDataFV (#18178)

moosebuild commented:
Job [Controlled App tests](https://civet.inl.gov/job/773316/) on 8dd0501 : invalidated by @rwcarlsen

moosebuild commented:
Job [Controlled App tests PETSc alt](https://civet.inl.gov/job/773319/) on 8dd0501 : invalidated by @rwcarlsen

lindsayad commented on framework/include/interfaces/Coupleable.h:
```diff
@@ -773,6 +773,16 @@ class Coupleable
+   * Second time derivative of a coupled variable for ad simulations
+   * @param var_name Name of coupled variable
+   * @param comp Component number for vector of coupled variables
+   * @return Reference to a VariableValue containing the second time derivative of the coupled
```
```suggestion
   * @return Reference to an ADVariableValue containing the second time derivative of the coupled
```

lindsayad commented on framework/include/variables/MooseVariableData.h:
```diff
@@ -315,6 +315,20 @@ class MooseVariableData
     return _ad_u_dot;
   }
 
+  const ADTemplateVariableValue<OutputType> & adUDotDot() const
+  {
+    _need_ad = _need_ad_u_dotdot = true;
+
+    if (!_time_integrator)
+      // If we don't have a time integrator (this will be the case for variables that are a part
+      // of the AuxiliarySystem) then we have no way to calculate _ad_u_dotdot and we are just
+      // going to copy the values from _u_dotdot. Of course in order to be able to do that we need
+      // to calculate _u_dotdot
+      _need_u_dotdot = true;
+
+    return _ad_u_dotdot;
+  }
```
I've been a guilty culprit in the past, but for things like this I'd prefer to keep the class definition clean of multi-line method definitions. You can put this outside the class definition as an `inline` function in the header

cticenhour replied:
Would you like me to move `adUDot` as well? That was here before I made these changes.

lindsayad replied:
You don't need to, but if you're pushing any more commits, feel free to

cticenhour replied:
Will do

cticenhour replied:
Do I need to be explicit about adding the `inline` to a method that isn't fully specialized? Take a look at 2fdf84b. The comments already present in `MooseVariableData.h` at the bottom of the file made that a little unclear.

lindsayad replied:
No you do not. I forgot that this was a templated class when I made my original comment. Any thing with a template parameter will be automatically `inline`

lindsayad commented on framework/include/variables/MooseVariableDataFV.h:
```diff
@@ -181,6 +181,20 @@ class MooseVariableDataFV
     return _ad_u_dot;
   }
 
+  const ADTemplateVariableValue<OutputType> & adUDotDot() const
+  {
+    _need_ad = _need_ad_u_dotdot = true;
+
+    if (!_time_integrator)
+      // If we don't have a time integrator (this will be the case for variables that are a part of
+      // the AuxiliarySystem) then we have no way to calculate _ad_u_dotdot and we are just going to
+      // copy the values from _u_dotdot. Of course in order to be able to do that we need to
+      // calculate _u_dotdot
+      _need_u_dotdot = true;
+
+    return _ad_u_dotdot;
+  }
```
same as above

lindsayad commented on framework/src/interfaces/Coupleable.C:
```diff
@@ -1776,6 +1776,23 @@ Coupleable::adCoupledDot(const std::string & var_name, unsigned int comp) const
+const ADVariableValue &
+Coupleable::adCoupledDotDot(const std::string & var_name, unsigned int comp) const
+{
+  const auto * var = getVarHelper<MooseVariableField<Real>>(var_name, comp);
```
```suggestion
  const auto * const var = getVarHelper<MooseVariableField<Real>>(var_name, comp);
```

lindsayad commented on test/tests/time_integrators/central-difference/ad_central_difference_dotdot.i:
```diff
@@ -0,0 +1,86 @@
+# correctly using the Central Difference method for an AD
+# variable.
+#
+# @Requirement F1.30
```
This is fancy. What does this mean?

cticenhour replied:
I am not entirely sure....precursor to the new SQA requirements I think, since this is copypasta from an old test. I will remove in the same commit that I switch to `inline` functions for `adUDotDot`

cticenhour replied:
Resolved in 2fdf84b

lindsayad commented on test/tests/time_integrators/central-difference/tests:
```diff
@@ -10,4 +9,12 @@
+    input = 'ad_central_difference_dotdot.i'
+    csvdiff = 'ad_central_difference_dotdot_out.csv'
+
+    requirement = "The CentralDifference time integrator shall correctly compute the second"
+                  " time derivative of an AD variable."
```
We should try to keep our requirements as implementation agnostic as possible, so probably something more like
```suggestion
    requirement = "A central differencing time integrator shall correctly compute the second"
                  " time derivative of an automatic differentiation variable."
```

lindsayad commented on test/tests/time_integrators/newmark-beta/tests:
```diff
@@ -20,4 +20,13 @@
+    input = 'ad_newmark_beta_dotdot.i'
+    csvdiff = 'ad_newmark_beta_dotdot_out.csv'
+
+    requirement = "The NewmarkBeta time integrator shall correctly compute the second"
+                  " time derivative of an AD variable using the default beta"
+                  " and gamma parameters."
```
```suggestion
    requirement = "A Newmark-beta method time integrator shall correctly compute the second"
                  " time derivative of an automatic differentiation variable using the default beta"
                  " and gamma parameters."
```

lindsayad reviewed (dismissed):
Overall looks good to me

cticenhour committed:
Address some review comments

Co-authored-by: Alex Lindsay <alexlindsay239@gmail.com>

cticenhour committed:
Address more review comments (#18178)

- Removes errant requirement number from new ad newmark and central 
difference tests. 
- Moved method definitions for adUDot and adUDotDot to outside class 
definition (still in header).

cticenhour committed:
Add finite volume AD dotdot test (#18178)

cticenhour commented:
@rwcarlsen I've defined the `adUDotDot` method for FV, as you requested. A new test for it is in 3690605, but I'm having an issue with it. I modeled this test after the new `ad_newmark_beta_dotdot.i` test I created in fa7291b, but it doesn't seem to work correctly. See the output below - the first three timesteps achieve the expected output, but then I cannot duplicate the rest of the series with pen-and-paper calcs using the variable value and dot. Is there an issue with the Postprocessor I'm using that produces the element average second time derivative? Not being well-versed on FV, that was my first thought. 

```
+----------------+----------------+
| time           | udotdot        |
+----------------+----------------+
|   0.000000e+00 |   0.000000e+00 |
|   1.000000e-01 |   0.000000e+00 |
|   2.000000e-01 |   4.761905e-02 |
|   3.000000e-01 |   9.070295e-02 |
|   4.000000e-01 |   3.444552e-02 |
|   5.000000e-01 |  -1.645405e-02 |
|   6.000000e-01 |  -1.488700e-02 |
+----------------+----------------+
```

rwcarlsen commented:
You could check the actual values from _udotdot[_qp] and see if they show what we expect (i.e. it reports the same value that the element average postprocessor is computing for its whole-elem integral).  For FV, we should be integrating over only a single qp - or at least - all qps should have copies of the single elemental value for an element.

cticenhour commented:
Alright, I've traced things back to the actual time-integrator code, and convinced myself it was working properly. I decided my test was a bit shoddy (based on the newmark AD test I also added here), so I replaced it with an FV simple diffusion transient example which works a lot better. I think this is ready to a final review.

cticenhour commented:
@lindsayad I also rebased this on top of your recent CNSFV changes due to conflicts, so double-check `MooseVariableDataFV` for anything I might have missed in the consolidation.

lindsayad reviewed (dismissed):
LGTM

cticenhour commented:
Had to rebase again due to a conflict created yesterday while the test boxes were down. Could I get another approval?
